Add extra logic in test_sfputil to make sure modules are restored

longhuan-cisco opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

Add extra logic (common for both CMIS and non-CMIS) in test_sfputil to make sure modules are restored fully after sfp reset, this is especially helpful for restoring CMIS modules which requires config/CMIS_state_machine to be replayed after sfp reset.

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311

### Approach
#### What is the motivation for this PR?

#### How did you do it?

#### How did you verify/test it?

Verified the testcase passed on setup which has 400G modules, and verified syslog the config/CMIS state machine was replayed properly after reset, modules were fully restored.
```
root@cmono-t0-dut:/home/cisco# show int trans info Ethernet8
Ethernet8: SFP EEPROM detected
...
        Application Advertisement: 400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400GBASE-DR4 (Cl 124) - Media Assign (0x1)
                                   100GAUI-2 C2M (Annex 135G) - Host Assign (0x55) - 100GBASE-DR (Cl 140) - Media Assign (0xf)
                                   200GAUI-4 C2M (Annex 120E) - Host Assign (0x11) - Undefined - Media Assign (0x5)
        CMIS Rev: 4.0
        Connector: MPO 1x12
        Encoding: N/A
        Extended Identifier: Power Class 4 (8.0W Max)
        Extended RateSelect Compliance: N/A
        Host Lane Count: 8
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
...
```
```
platform_tests/sfp/test_sfputil.py::test_check_sfputil_reset[cmono-t0-dut-None] PASSED                                  [100%]

====================================================== warnings summary =======================================================
../../usr/local/lib/python3.8/dist-packages/paramiko/transport.py:236
  /usr/local/lib/python3.8/dist-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
    "class": algorithms.Blowfish,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------------------------- generated xml file: /data/tests/logs/tr_2024-06-03-18-13-32.xml -------------------------------
========================================== 1 passed, 1 warning in 1434.78s (0:23:54) ==========================================
INFO:root:Can not get Allure report URL. Please check logs
```

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/platform_tests/sfp/test_sfputil.py:208:13: E125 continuation line with same indent as next logical line<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

prgeor commented on tests/platform_tests/sfp/conftest.py:
```diff
@@ -77,3 +77,35 @@ def stop_xcvrd(duthost):
     restart_pmon(duthost)
+
+
+class StopXcvrd:
```
@longhuan-cisco no need to stop the Xcvrd. Latest changes in Xcvrd ensures that the DOM is not polled if the port is 'admin' down. @mihirpat1 vis

longhuan-cisco replied:
As we discussed in the email, Mihir’s PR [#449](https://github.com/sonic-net/sonic-platform-daemons/pull/449) will only make sure CMIS module’s polling is disabled during interface shut stage. Non-CMIS modules will be always treated as CMIS_STATE_READY after insertion, which causes is_port_dom_monitoring_disabled() to be False no matter whether interface is shut or not.
So to cover both CMIS/non-CMIS modules, we would still need to stop_xcvrd during sfp_reset_cmd, and restart_xcvrd after it.

prgeor replied:
@longhuan-cisco can you disable/enable DOM via:-

`config interface transceiver dom PORT_NAME (enable|disable) `

before and after test

longhuan-cisco replied:
@prgeor sure, I updated the diff by using `config interface transceiver dom` to disable/enable dom. Please check.

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/platform_tests/sfp/test_sfputil.py:18:1: F811 redefinition of unused 'wait_until' from line 16<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -100,27 +207,61 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
+    # to get config/state machine/etc replayed, so that the modules can be fully
+    # restored.
+    logging.info("Shutdown ports before sfp reset")
+    shutdown_result = duthost.shell_cmds(cmds=cmds_down)
```
@longhuan-cisco Can we use a context manager to toggle the port in shut/startup state as well as toggle lpmode.
The goal here is if any exception occurs, the context manager will ensure that the port is taken out of shut state/lpmode before returning/proceeding to next port.

Following will be the template for it
```
with port_disabled:
    sfputil_reset()
    sleep(5)
    toggle_lpmode_if_required()
verify_link_is_restored()
```
https://www.geeksforgeeks.org/context-manager-in-python/

longhuan-cisco replied:
good suggestion. updated to context manager.

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -100,27 +207,61 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
+    lpmode_show = duthost.command(cmd_sfp_show_lpmode)
+    original_lpmode = parse_output(lpmode_show["stdout_lines"][2:])
+
+    logging.info("Disable DOM polling to avoid race condition during sfp reset")
```
@longhuan-cisco Can we disable DOM monitoring before issuing shutdown command?

longhuan-cisco replied:
Moved it to before shutdown

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -100,27 +207,61 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
+        logging.info("resetting {} physical interface {}".format(intf, phy_intf))
+        reset_result = duthost.command(cmd_sfp_reset_intf)
+        assert reset_result["rc"] == 0, "'{}' failed".format(cmd_sfp_reset_intf)
+        time.sleep(5)
```
@longhuan-cisco Can you please name this 5s delay to `I2C_WAIT_TIME_AFTER_RESET` or something meaningful?

longhuan-cisco replied:
Renamed it

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -100,27 +207,61 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
+        reset_result = duthost.command(cmd_sfp_reset_intf)
+        assert reset_result["rc"] == 0, "'{}' failed".format(cmd_sfp_reset_intf)
+        time.sleep(5)
     sleep_time = 60
```
Is this this sleep still needed in addition to the above 5s sleep? Wondering since we are already polling for the link to be up as part of interface startup command.

longhuan-cisco replied:
Removed the unnecessary sleep since we already have wait_util/polling logic

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -100,27 +207,61 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
-            reset_result = duthost.command("{} {}".format(cmd_sfp_reset, intf))
-            assert reset_result["rc"] == 0, "'{} {}' failed".format(cmd_sfp_reset, intf)
-            time.sleep(5)
+    wait_after_restore_lpmode = 10
```
@longhuan-cisco Can you please share the reason for 10s delay instead of a shorter time delay here?

longhuan-cisco replied:
changed to shorter delay

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -49,6 +57,105 @@ def test_check_sfputil_presence(duthosts, enum_rand_one_per_hwsku_frontend_hostn
+    """Check and return the down ports among the given ports."""
+    ports_down = duthost.interface_facts(up_ports=ports)["ansible_facts"][
+        "ansible_interface_link_down_ports"]
+    db_ports_down = duthost.show_interface(command="status", up_ports=ports)["ansible_facts"][
```
@longhuan-cisco Can you please explain the difference between ports_down and db_ports_down (wondering why db_ports_down is not sufficient)?

longhuan-cisco replied:
Updated to only rely on db_ports_down

wsycqyz commented:
Hi @mihirpat1 , there are some questions left in the PR, can you have a look?

mihirpat1 commented:
> Hi @mihirpat1 , there are some questions left in the PR, can you have a look?

@wsycqyz Can you please share the open questions for me?

longhuan-cisco commented:
The below two pre-commit check failures are due to baseline, because they started to happen right after merging master to local branch https://github.com/sonic-net/sonic-mgmt/pull/13108/commits/b0dd014f6bba79df0a9384e703499b1907b4656b:
https://github.com/sonic-net/sonic-mgmt/pull/13108/checks?check_run_id=27298781282
https://github.com/sonic-net/sonic-mgmt/pull/13108/checks?check_run_id=27298780950

jamesan47 commented:
@kevinwangsk please help with review / approval.  Thanks!

mihirpat1 reviewed (commented):
@longhuan-cisco Can you please test the changes on a device with breakout cable and ensure all subports are linked up after one of the subport has gone through reset.
Also, please ensure the DOM config is disabled only for the first subport of a breakout group.

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -122,54 +326,46 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
-        "Some interfaces are down: {}".format(intf_facts["ansible_interface_link_down_ports"])
+            cmd_sfp_reset_intf = "{} {}".format(cmd_sfp_reset, intf)
+            logging.info("resetting {} physical interface {}".format(intf, phy_intf))
+            reset_result = duthost.command(cmd_sfp_reset_intf)
```
@longhuan-cisco In case of breakout ports, shouldn't we issue reset on one of the subports in breakout group and issue shut/no shut for other subports in the corresponding breakout group?

longhuan-cisco replied:
agree, updated as the same. thanks

longhuan-cisco commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -15,12 +15,22 @@
+cmd_dom_disable = "disable"
+cmd_dom_enable = "enable"
+
+I2C_WAIT_TIME_AFTER_SFP_RESET = 3  # in seconds
```
updated to 5s

longhuan-cisco commented:
> @longhuan-cisco Can you please test the changes on a device with breakout cable and ensure all subports are linked up after one of the subport has gone through reset. Also, please ensure that the DOM config is disabled only for the first subport of a breakout group.

@mihirpat1  Updated code to only disable DOM for first subport  of breakout group, and tested it.

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -30,6 +39,144 @@
+        self.original_lpmode_state = None
+        self.skip_dom_polling_handle = skip_dom_polling_handle
+        self.skip_lpmode_handle = skip_lpmode_handle
+        self.wait_after_dom_config = 1
```
Can we change this to 5?

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -30,6 +39,144 @@
+        self.phy_intf = phy_intf
+        self.original_lpmode_state = None
+        self.skip_dom_polling_handle = skip_dom_polling_handle
+        self.skip_lpmode_handle = skip_lpmode_handle
```
@longuan-cisco We need to skip lpmode config for all CMIS cables as well as cables which are `Power Class 1 Module`. Can you please handle this scenario?

longhuan-cisco replied:
Added the handling for both

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -40,6 +187,98 @@ def check_interfaces_up(duthost, namespace, up_ports):
+        phy_intfs_to_test_per_asic: dict of all physical interfaces to test (key: physical port number,
+                                    value: list of logical interfaces under this physical port)
+    """
+    portmap, dev_conn = get_dev_conn(duthost,
```
@longhuan-cisco Based on my understanding, the portmap here does not yield physical port numbers for device with breakout port. `portmap` here seems to be a dict of logical_port_name as key and an integer number starting from 0 with all logical ports in a sorted order.
https://github.com/sonic-net/sonic-mgmt/blob/e7f3a300c9f476ddc1d3a2fe51318b33d226f0ec/ansible/library/config_facts.py#L99-L100

I also tried executing your changes on a device with breakout ports and observed that every physical port has just 1 logical port. Are you observing similar behavior on a device with breakout ports too?

Due to the above issue, the DOM config command will fail for logical ports which have subport > 1.

longhuan-cisco replied:
Right, today's portmap is not reliable for breakout case. 
Updated it to rely on get_physical_port_indices() function which is based on `index` from port config.

```
platform_tests/sfp/test_sfputil.py::test_check_sfputil_reset[mathilda-01-None] PASSED                              [100%]

  Ethernet8                 2320      25G   9100    N/A    etp2a  routed      up       up  QSFP28 or later         N/A
  Ethernet9                 2321      25G   9100    N/A    etp2b  routed      up       up  QSFP28 or later         N/A
 Ethernet10                 2322      25G   9100    N/A    etp2c  routed      up       up  QSFP28 or later         N/A
 Ethernet11                 2323      25G   9100    N/A    etp2d  routed      up       up  QSFP28 or later         N/A

09/09/2024 03:54:02 test_sfputil.get_phy_intfs_to_test_per_a L0304 INFO   | Interfaces to test for asic None: {2: [u'Ethernet8', u'Ethernet9', u'Ethernet10', u'Ethernet11']}

09/09/2024 03:54:02 test_sfputil.skip_lpmode_handle_for_phys L0176 INFO   | Skip lpmode handling for physical interface 2 as it's CMIS compliant(False) or Power Class 1(True)

09/09/2024 03:54:02 test_sfputil.disable                     L0080 INFO   | Disable DOM polling to avoid race condition during sfp reset for Ethernet8
09/09/2024 03:54:08 test_sfputil.disable                     L0089 INFO   | Shutdown Ethernet8 before sfp reset
09/09/2024 03:54:10 test_sfputil.disable                     L0089 INFO   | Shutdown Ethernet9 before sfp reset
09/09/2024 03:54:13 test_sfputil.disable                     L0089 INFO   | Shutdown Ethernet10 before sfp reset
09/09/2024 03:54:15 test_sfputil.disable                     L0089 INFO   | Shutdown Ethernet11 before sfp reset
09/09/2024 03:54:35 test_sfputil.restore                     L0124 INFO   | Startup Ethernet8 after sfp reset to restore module
09/09/2024 03:54:37 test_sfputil.restore                     L0130 INFO   | Restore DOM polling after sfp reset for Ethernet8
09/09/2024 03:54:38 test_sfputil.restore                     L0124 INFO   | Startup Ethernet9 after sfp reset to restore module
09/09/2024 03:54:40 test_sfputil.restore                     L0124 INFO   | Startup Ethernet10 after sfp reset to restore module
09/09/2024 03:54:42 test_sfputil.restore                     L0124 INFO   | Startup Ethernet11 after sfp reset to restore module
```

longhuan-cisco commented:
The pre-commit sanity check failure [Azure.sonic-mgmt (Test onboarding t1 testcases by Elastictest - optional) ](https://github.com/sonic-net/sonic-mgmt/pull/13108/checks?check_run_id=29855101954) seems to be some baseline issue, also saw it on other PRs':
https://github.com/sonic-net/sonic-mgmt/pull/14474/checks?check_run_id=29863617790
https://github.com/sonic-net/sonic-mgmt/pull/14472/checks?check_run_id=29853135820

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -40,6 +208,97 @@ def check_interfaces_up(duthost, namespace, up_ports):
+            wait_time = port_down_wait_time
+
+    logging.info("Wait for ports to come {}: {}".format(expect_status_str, ports))
+    is_ok = wait_until(wait_time, 5, 0,
```
@longhuan-cisco Can we change the polling interval from 5s to 1s to optimize test case run time?

longhuan-cisco replied:
updated

mihirpat1 reviewed (commented):
@prgeor for viz
@longhuan-cisco I am seeing that after executing the `test_check_sfputil_reset` testcase individually, a config reload is being triggered due to a failure seen while executing `core_dump_and_config_check`. Are you observing a similar behavior too?

Based on my understanding, this config reload is being triggered due to [config_db_check_pass](https://github.com/sonic-net/sonic-mgmt/blob/9e3cc9b8394adf8ce9fa2215adec4f17d0cb7e27/tests/conftest.py#L2340) being set to False.
This is because the current running_config has `"dom_polling":"enabled"` key-value introduced to PORT table of CONFIG_DB since the testcase disables and enables DOM.
By default, transceiver DOM polling is enabled but the `dom_polling` field may not be present in PORT table of CONFIG_DB.
Hence, can we ensure that the config reload is not triggered if
1. pre-running config does not have `dom_polling` field present
2. AND the `dom_polling` field is set to `enabled` in running-config post testcase execution

prgeor reviewed (commented):
@Junchao-Mellanox @keboliu can you please review?

longhuan-cisco commented:
> @prgeor for viz @longhuan-cisco I am seeing that after executing the `test_check_sfputil_reset` testcase individually, a config reload is being triggered due to a failure seen while executing `core_dump_and_config_check`. Are you observing a similar behavior too?
> 
> Based on my understanding, this config reload is being triggered due to [config_db_check_pass](https://github.com/sonic-net/sonic-mgmt/blob/9e3cc9b8394adf8ce9fa2215adec4f17d0cb7e27/tests/conftest.py#L2340) being set to False. This is because the current running_config has `"dom_polling":"enabled"` key-value introduced to PORT table of CONFIG_DB since the testcase disables and enables DOM. By default, transceiver DOM polling is enabled but the `dom_polling` field may not be present in PORT table of CONFIG_DB. 

@mihirpat1  Yes, I observed the same: testcase passed, but there's a WARNING on config_check due to that `dom_polling=enabled` added after restoring the port, and config reload is triggered as a result.

> Hence, can we ensure that the config reload is not triggered if
> 
> 1. pre-running config does not have `dom_polling` field present
> 2. AND the `dom_polling` field is set to `enabled` in running-config post testcase execution

To avoid the config reload:
Were you suggesting to change the logic in conftest.py:core_dump_and_config_check() to exclude checking `dom_polling` field, if there's no `dom_polling` present in port config pre-test AND there's `dom_polling=enabled` in port config post-test?
yea, this would work.

Alternatively, simpler approach might be (although a little bit "raw"): we may clear this filed by `sonic-db-cli  CONFIG_DB HDEL 'PORT|Ethernet0' 'dom_polling'` in the end for those relevant interfaces.  
what do you think?

mihirpat1 commented:
> > @prgeor for viz @longhuan-cisco I am seeing that after executing the `test_check_sfputil_reset` testcase individually, a config reload is being triggered due to a failure seen while executing `core_dump_and_config_check`. Are you observing a similar behavior too?
> > Based on my understanding, this config reload is being triggered due to [config_db_check_pass](https://github.com/sonic-net/sonic-mgmt/blob/9e3cc9b8394adf8ce9fa2215adec4f17d0cb7e27/tests/conftest.py#L2340) being set to False. This is because the current running_config has `"dom_polling":"enabled"` key-value introduced to PORT table of CONFIG_DB since the testcase disables and enables DOM. By default, transceiver DOM polling is enabled but the `dom_polling` field may not be present in PORT table of CONFIG_DB.
> 
> @mihirpat1 Yes, I observed the same: testcase passed, but there's a WARNING on config_check due to that `dom_polling=enabled` added after restoring the port, and config reload is triggered as a result.
> 
> > Hence, can we ensure that the config reload is not triggered if
> > 
> > 1. pre-running config does not have `dom_polling` field present
> > 2. AND the `dom_polling` field is set to `enabled` in running-config post testcase execution
> 
> To avoid the config reload: Were you suggesting to change the logic in conftest.py:core_dump_and_config_check() to exclude checking `dom_polling` field, if there's no `dom_polling` present in port config pre-test AND there's `dom_polling=enabled` in port config post-test? yea, this would work.
> 
> Alternatively, simpler approach might be (although a little bit "raw"): we may clear this filed by `sonic-db-cli CONFIG_DB HDEL 'PORT|Ethernet0' 'dom_polling'` in the end for those relevant interfaces. what do you think?

@longhuan-cisco I think fixing in [this](https://github.com/sonic-net/sonic-mgmt/blob/9e3cc9b8394adf8ce9fa2215adec4f17d0cb7e27/tests/conftest.py#L2300) section will be more gene

> > @prgeor for viz @longhuan-cisco I am seeing that after executing the `test_check_sfputil_reset` testcase individually, a config reload is being triggered due to a failure seen while executing `core_dump_and_config_check`. Are you observing a similar behavior too?
> > Based on my understanding, this config reload is being triggered due to [config_db_check_pass](https://github.com/sonic-net/sonic-mgmt/blob/9e3cc9b8394adf8ce9fa2215adec4f17d0cb7e27/tests/conftest.py#L2340) being set to False. This is because the current running_config has `"dom_polling":"enabled"` key-value introduced to PORT table of CONFIG_DB since the testcase disables and enables DOM. By default, transceiver DOM polling is enabled but the `dom_polling` field may not be present in PORT table of CONFIG_DB.
> 
> @mihirpat1 Yes, I observed the same: testcase passed, but there's a WARNING on config_check due to that `dom_polling=enabled` added after restoring the port, and config reload is triggered as a result.
> 
> > Hence, can we ensure that the config reload is not triggered if
> > 
> > 1. pre-running config does not have `dom_polling` field present
> > 2. AND the `dom_polling` field is set to `enabled` in running-config post testcase execution
> 
> To avoid the config reload: Were you suggesting to change the logic in conftest.py:core_dump_and_config_check() to exclude checking `dom_polling` field, if there's no `dom_polling` present in port config pre-test AND there's `dom_polling=enabled` in port config post-test? yea, this would work.
> 
> Alternatively, simpler approach might be (although a little bit "raw"): we may clear this filed by `sonic-db-cli CONFIG_DB HDEL 'PORT|Ethernet0' 'dom_polling'` in the end for those relevant interfaces. what do you think?

@longhuan-cisco We can proceed with the alternative aproach (clearing the `dom_polling` field after enabling DOM)

Junchao-Mellanox commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -30,6 +44,183 @@
+            # Restore lpmode if needed, because sfp reset will reset lpmode to default
+            logging.info("Check output of '{}' after sfp reset".format(self.cmd_sfp_show_lpmode))
+            lpmode_show = self.duthost.command(self.cmd_sfp_show_lpmode)
+            after_lpmode_state = \
```
why do we need recover LPM here? I think LPM should be handled by xcvrd.

longhuan-cisco replied:
For non-cmis module, xcvrd doesn't handle LPM, which might need to be restored after sfputil reset

Junchao-Mellanox commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -30,6 +44,183 @@
+        logging.info("Startup {} after sfp reset to restore module".format(self.logical_intf))
+        startup_result = self.duthost.command(self.cmd_up)
+        assert startup_result["rc"] == 0, "Startup {} failed".format(self.logical_intf)
+        assert check_interface_status(self.duthost, [self.logical_intf], expect_up=True)
```
do we need a per port check here? I see there is a overall check at line 446

longhuan-cisco replied:
per-port check is to make sure each module is in expected state at each step when testing the current port. 
overall check is to double check just in case, e.g. if any not-currently-being-tested port gets impacted and goes down as collateral.

Junchao-Mellanox commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -30,6 +44,183 @@
+        is_cmis = "cmis_rev" in xcvr_info_output
+        is_power_class_1 = "Power Class 1" in xcvr_info_output
+
+        if is_cmis or is_power_class_1:
```
could you please add comment about why cmis is skipped?

mihirpat1 replied:
@longhuan-cisco Let's not skip lpmode toggling for CMIS modules.
Based on my analysis, the toggling is needed since some platforms may not have LowPwrRequestHW signal ASSERTED. Hence, triggering `sfputil reset` will not cause any change to Module State for such platforms and will eventually require user to toggle the lpmode manually.

longhuan-cisco replied:
> could you please add comment about why cmis is skipped?

added

longhuan-cisco replied:
> @longhuan-cisco Let's not skip lpmode toggling for CMIS modules. Based on my analysis, the toggling is needed since some platforms may not have LowPwrRequestHW signal ASSERTED. Hence, triggering `sfputil reset` will not cause any change to Module State for such platforms and will eventually require user to toggle the lpmode manually.

@mihirpat1 

<img width="1008" alt="image" src="https://github.com/user-attachments/assets/23c4bf66-a66b-45cc-ac67-5c457688fb04">
Doesn't cmis_mgr always turns CMIS module to high power (via `cmis_api.set_lpmode(False)`)  in CMIS state machine when starting up the interface? And set_lpmode forces transition from LowPwr to HighPower state under SW control regardless of LowPwrRequestHW.

https://github.com/sonic-net/sonic-platform-common/blob/daeed65100c9250a704de223a771fdc7416a90f0/sonic_platform_base/sonic_xcvr/api/public/cmis.py#L1089

thus, I didn't follow why CMIS module still needs manual toggle (`sfputil lpmode`) after reset.

Junchao-Mellanox commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -122,54 +404,50 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
+                                                                enum_frontend_asic_index,
+                                                                xcvr_skip_list)
+    for phy_intf, logical_intfs_list in phy_intfs_to_test_per_asic.items():
+        logical_intf = logical_intfs_list[0]
```
only test the first logic port? would it loss coverage?

longhuan-cisco replied:
Only reset the first logical interface under the physical interface, since sfputil command acts on physical port fully. 
added a comment in code to explain this.

mihirpat1 commented on tests/platform_tests/sfp/test_sfputil.py:
```diff
@@ -122,54 +404,50 @@ def test_check_sfputil_eeprom(duthosts, enum_rand_one_per_hwsku_frontend_hostnam
+                "Interface presence is not 'Present' for {}".format(logical_intf)
+
+    # Check interface status for all interfaces in the end just in case
+    assert check_interface_status(duthost,
```
@longhuan-cisco Can we deinitialize all lanes (similar to what we do via set_datapath_deinit API) after issuing sfputil reset?
This is needed to ensure that when the module transitions to high power mode, the modules does not initialize itself with the default settings. Hence, this will give software the control of initializing the module after reset is issued and prevent ports from linking up just after module is brought in high power mode.

longhuan-cisco replied:
Were you talking about adding something like set_datapath_deinit on cmis modules right after reset, to avoid the potential early link-up (on certain module) due to manually toggling to high power mode after reset?

Since there's still question on whether to manually toggle lpmode after reset for cmis, will wait after closure on discussion https://github.com/sonic-net/sonic-mgmt/pull/13108/files#r1785319428

mihirpat1 replied:
@longhuan-cisco I have responded to the comment now. Let me know if you still have further questions.

longhuan-cisco replied:
Updated as discussed over email.

prgeor commented:
@mihirpat1 can you review

mihirpat1 commented:
> @mihirpat1 can you review

@longhuan-cisco Can you please help in addressing the below per our discussion?
https://github.com/sonic-net/sonic-mgmt/pull/13108#discussion_r1785387333

longhuan-cisco commented:
> @longhuan-cisco Can you please help in addressing the below per our discussion? [#13108 (comment)](https://github.com/sonic-net/sonic-mgmt/pull/13108#discussion_r1785387333)

@mihirpat1 
I addressed as what we discussed. 

And I also added logic to skip `config interface shudown/startup` for already admin-down ports, because of two reasons: 1) no need to shutdown already admin-down ports; 2) to avoid unexpectedly startup pre-test admin-down port, which can cause config check failed. This skip is especially needed, on certain test topologies/configurations, where some of the ports in admin-down state always:
```
     Interface                                    Lanes    Speed    MTU    FEC    Alias            Vlan    Oper    Admin                                             Type    Asym PFC
--------------  ---------------------------------------  -------  -----  -----  -------  --------------  ------  -------  -----------------------------------------------  ----------
     Ethernet0  2304,2305,2306,2307,2308,2309,2310,2311     400G   9100    N/A     etp0          routed    down     down  QSFP-DD Double Density 8X Pluggable Transceiver         off
     Ethernet8  2320,2321,2322,2323,2324,2325,2326,2327     400G   9100    N/A     etp1           trunk      up       up  QSFP-DD Double Density 8X Pluggable Transceiver         off
    Ethernet16  2312,2313,2314,2315,2316,2317,2318,2319     400G   9100    N/A     etp2           trunk      up       up  QSFP-DD Double Density 8X Pluggable Transceiver         off
...
   Ethernet184                          0,1,2,3,4,5,6,7     400G   9100    N/A    etp23           trunk      up       up  QSFP-DD Double Density 8X Pluggable Transceiver         off
   Ethernet192          256,257,258,259,260,261,262,263     400G   9100    N/A    etp24           trunk      up       up  QSFP-DD Double Density 8X Pluggable Transceiver         off
   Ethernet200                    8,9,10,11,12,13,14,15     400G   9100    N/A    etp25          routed    down     down  QSFP-DD Double Density 8X Pluggable Transceiver         off
   Ethernet208  1024,1025,1026,1027,1028,1029,1030,1031     400G   9100    N/A    etp26          routed    down     down  QSFP-DD Double Density 8X Pluggable Transceiver         off
   Ethernet216          768,769,770,771,772,773,774,775     400G   9100    N/A    etp27          routed    down     down  QSFP-DD Double Density 8X Pluggable Transceiver         off
...
```

bmridul commented:
@prgeor, Mihir has reviewed the PR. Can u pls merge this.

kevinskwang commented:
@Junchao-Mellanox could you resolve the conversation if your comments are addressed?

mssonicbld commented:
@longhuan-cisco PR conflicts with 202405 branch

kevinskwang commented:
@longhuan-cisco pls address the conflicts on 202405.

longhuan-cisco commented:
@kevinskwang I've addressed conflicts and raised cherry-pick PR already https://github.com/sonic-net/sonic-mgmt/pull/15543

eyakubch commented:
@longhuan-cisco Hi, we need this changes also for 202311 branch, can you please also make PR for 202311 ?
